### PR TITLE
FORMS-211 # pristine/dirty attributes and CSS classes to signal user activity

### PR DESCRIPTION
**Pull request by simonmarklar**
- Added `.bm-pristine` and `.bm-dirty` classes to forms, subforms and elements
- Added Mixin so the code can be used in various models

### CODE REVIEW

**Comment by jokeyrhyme** [comment]
In FORMS-194, we'll be added `bm-form` and `bm-formelement` BEM CSS classes.
In light of that, I wonder if we should be using the following here:
- `bm-form-pristine`
- `bm-form-dirty`
- `bm-formelement-pristine`
- `bm-formelement-dirty`

This follows the convention we found and discussed: https://github.com/blinkmobile/docs/wiki/Code-Style:-Cascading%20Style%20Sheets

> 2 hyphens denote a modifier state

**Comment by jokeyrhyme** [comment]
Other than that, it looks good to me. I like that you updated the `defaults` hash in the Backbone constructors, both for documentation and engine optimisation.

**Comment by jokeyrhyme** [comment]
I also like that you've gone with `is...` for Boolean properties.

**Comment by simonmarklar** [comment]
having a separate form and element class could come in handy. I'll add it in.

**Comment by jokeyrhyme** [comment]
LGTM :+1: 
Can we get another coder and a customer advocate to take a look?

**Comment by simonmarklar** [comment]
yeah i want to run it past sam as he was asking for it, but he ran away this morning before i could grab him

Its important to note that this only adds/removes the classes. And of course when I tried to make it so that only pristine elements run a dirty check everything exploded (I traced it back to an initialization/render/behavior order problem, slowly backed away from that one, its like a house of cards and way out of scope for this ticket)

its easy to hide all errors until the form or element is dirty now, so i hope that does what sam needs.

**Comment by jokeyrhyme** [comment]
Oh, it's be great if you updated the CSS documentation, too:
https://github.com/blinkmobile/forms/blob/develop/docs/css-selectors.md

**Comment by simonmarklar** [comment]
Documentation is updated, I also changed the previous invalid class add/change to work in the same way. Might as well be consistent :)

**Comment by jokeyrhyme** [comment]
LGTM :+1:

**Comment by simonmarklar** [comment]
i've run this past sam and he's happy with it

**Comment by jokeyrhyme** [comment]
LGTM :+1:

File: forms/jqm/element.js [hunk 1]
```diff
@@ -158,18 +177,12 @@ define(function (require) {
               text = formsErrors.toErrorString(error);
             }
 
-            return $(document.createElement('li')).text(text).addClass('bm-errors__bm-listitem');
+            return $(document.createElement('li')).text(text).addClass('bm-errors__bm-listitem' + requiredClass);
```
Comment (ashishtilara) [inline forms/jqm/element.js:180]: should there be a space, `'bm-errors__bm-listitem ' + requiredClass` ?

Comment (ashishtilara) [inline forms/jqm/element.js:180]: oh never-mind, saw the requiredClass adds that.

**Comment by jokeyrhyme** [comment]
Apparently there are some merge conflicts that need to be addressed now, too. :(

File: forms/models/elements/heading.js [hunk 2]
```diff
@@ -32,9 +40,11 @@ define(['forms/models/element'], function (Element) {
           self.attributes.level = headingType + 1;
         }
       }
+      ElementModel.prototype.initialize.apply(this, arguments);
+    },
 
-      Element.prototype.initialize.apply(this, arguments);
-    }
+    setDirty: $.noop,
+    setPristine: $.noop
```
Comment (ashishtilara) [inline forms/models/elements/heading.js:47]: is it working adding the same to button fields?!

Comment (simonmarklar) [inline forms/models/elements/heading.js:47]: do you mean 

> is it working adding the same to button fields?!

 ??
If so, probably, i'll add it in

**Comment by simonmarklar** [comment]
@ashishtilara 
- the view needs some way to add remove the class. The _changed property doesnt broadcast any events - its just for reference. Without listening for an event, I cant see how the view would know to check the changed property and adjust the class. This use case is exactly what backbone events were made for :

http://backbonejs.org/#View

> The general idea is to organize your interface into logical views, backed by models, each of which can be updated independently when the model changes, without having to redraw the page. Instead of digging into a JSON object, looking up an element in the DOM, and updating the HTML by hand, you can bind your view's render function to the model's "change" event — and now everywhere that model data is displayed in the UI, it is always immediately up to date.

in backbone you're supposed to listen to change and then re-render the entire view, I dont think jQuery mobile would like it if we did that :O
- Because the dirty class is tied to the `change:value` event the conditional logic fields will also be marked as dirty. I think this is the behavior that we want - Discussing it with Sam, he agreed that any value set by an API should be considered dirty and I think that applies here.
- I include the whole thing because I want it to match what the server  sends down, and at the time of writing it i had no idea what was necessary and what wasn't . But yeah I agree it is getting quite huge. i'll trim it down

**Comment by jokeyrhyme** [comment]
I actually think it's terrific to be using real-world definitions, regardless of the size.

**Comment by ashishtilara** [comment]
Okay that makes sense...

I also agree with using real-world definitions, and what I am asking is we should only include the definitions that are used by test since this is open-source, someone outside the blink tries to work on it, and sees too many definitions will find it hard to understand. Chances of confusion increases as well but thats me. If you guys think its not confusing, rest looks good to me.

**Comment by jokeyrhyme** [comment]
@ashishtilara that's a great point.

I agree that for feature development and bug fixes it is easier for someone new to the project to be working with a minimal definition.

Maybe in future we should separate / clearly-mark the real-world definitions. There should always be a reduced minimal definition that accompanies new work, even if there is a separate test that includes a real-world definition.

**Comment by jokeyrhyme** [comment]
So, where are we up to here? @ashishtilara do you have any unanswered questions? Does this have your seal of approval once @simonmarklar sorts out the merge conflicts?

**Comment by ashishtilara** [comment]
LGTM :+1:

**Comment by simonmarklar** [comment]
cherry picked version of this branch has been merged. closing.

**Pull request closed without merging**